### PR TITLE
Add Zscaler IaC Scan to Code Scanning

### PR DESCRIPTION
**Pull request by mvecha**
## Pre-requisites

- [x] Prior to submitting a new workflow, please apply to join the GitHub Technology Partner Program: [partner.github.com/apply](https://partner.github.com/apply?partnershipType=Technology+Partner).

---

### **Please note that at this time we are only accepting new starter workflows for Code Scanning. Updates to existing starter workflows are fine.**

---

## Tasks

**For _all_ workflows, the workflow:**

- [x] Should be contained in a `.yml` file with the language or platform as its filename, in lower, [_kebab-cased_](https://en.wikipedia.org/wiki/Kebab_case) format (for example, [`docker-image.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-image.yml)).  Special characters should be removed or replaced with words as appropriate (for example, "dotnet" instead of ".NET").
- [x] Should use sentence case for the names of workflows and steps (for example, "Run tests").
- [x] Should be named _only_ by the name of the language or platform (for example, "Go", not "Go CI" or "Go Build").
- [x] Should include comments in the workflow for any parts that are not obvious or could use clarification.
- [x] Should specify least priviledge [permissions](https://docs.github.com/en/actions/security-guides/automatic-token-authentication#modifying-the-permissions-for-the-github_token) for `GITHUB_TOKEN` so that the workflow runs successfully. 

**For _CI_ workflows, the workflow:**

- [ ] Should be preserved under [the `ci` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [ ] Should include a matching `ci/properties/*.properties.json` file (for example, [`ci/properties/docker-publish.properties.json`](https://github.com/actions/starter-workflows/blob/main/ci/properties/docker-publish.properties.json)).
- [ ] Should run on `push` to `branches: [ $default-branch ]` and `pull_request` to `branches: [ $default-branch ]`.
- [ ] Packaging workflows should run on `release` with `types: [ created ]`.
- [ ] Publishing workflows should have a filename that is the name of the language or platform, in lower case, followed by "-publish" (for example, [`docker-publish.yml`](https://github.com/actions/starter-workflows/blob/main/ci/docker-publish.yml)).

**For _Code Scanning_ workflows, the workflow:**

- [x] Should be preserved under [the `code-scanning` directory](https://github.com/actions/starter-workflows/tree/main/ci).
- [x] Should include a matching `code-scanning/properties/*.properties.json` file (for example, [`code-scanning/properties/codeql.properties.json`](https://github.com/actions/starter-workflows/blob/main/code-scanning/properties/codeql.properties.json)), with properties set as follows:
  - [x] `name`: Name of the Code Scanning integration.
  - [x] `organization`: Name of the organization producing the Code Scanning integration.
  - [x] `description`: Short description of the Code Scanning integration.
  - [x] `categories`: Array of languages supported by the Code Scanning integration.
  - [x] `iconName`: Name of the SVG logo representing the Code Scanning integration. This SVG logo must be present in [the `icons` directory](https://github.com/actions/starter-workflows/tree/main/icons).
- [x] Should run on `push` to `branches: [ $default-branch, $protected-branches ]` and `pull_request` to `branches: [ $default-branch ]`. We also recommend a `schedule` trigger of `cron: $cron-weekly` (for example, [`codeql.yml`](https://github.com/actions/starter-workflows/blob/c59b62dee0eae1f9f368b7011cf05c2fc42cf084/code-scanning/codeql.yml#L14-L21)).

**Some general notes:**

- [ ] This workflow must _only_ use actions that are produced by GitHub, [in the `actions` organization](https://github.com/actions), **or**
- [ ] This workflow must _only_ use actions that are produced by the language or ecosystem that the workflow supports.  These actions must be [published to the GitHub Marketplace](https://github.com/marketplace?type=actions).  We require that these actions be referenced using the full 40 character hash of the action's commit instead of a tag.  Additionally, workflows must include the following comment at the top of the workflow file:
    ```
    # This workflow uses actions that are not certified by GitHub.
    # They are provided by a third-party and are governed by
    # separate terms of service, privacy policy, and support
    # documentation.
    ```
- [ ] Automation and CI workflows should not send data to any 3rd party service except for the purposes of installing dependencies.
- [ ] Automation and CI workflows cannot be dependent on a paid service or product.

### CODE REVIEW

File: code-scanning/zscaler-iac-scan.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    - cron: $cron-weekly
+
+jobs:
+  zscaler-iac-scan:
+    runs-on: ubuntu-latest
+    steps:
+      - name : Code Checkout
+        uses: actions/checkout@v2
+      - name : Zscaler IAC Scan
+        uses : ZscalerCWP/Zscaler-IaC-Action@v0.2.0
```
Comment (Phantsure) [inline code-scanning/zscaler-iac-scan.yml:26]: Can you please change `v0.2.0` to the respective commit sha? Thanks

File: code-scanning/zscaler-iac-scan.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+name: Zscaler IaC Scan
+on:
+  push:
+    branches: [ $default-branch, $protected-branches ]
+  pull_request:
+    branches: [ $default-branch ]
+  schedule:
+    - cron: $cron-weekly
+
```
Comment (Phantsure) [inline code-scanning/zscaler-iac-scan.yml:18]: Can you add permissions? Look here https://github.com/actions/starter-workflows/pull/1480/files

**Review by Phantsure — CHANGES_REQUESTED** [review]
Kindly have a look at my comments

File: code-scanning/zscaler-iac-scan.yml (hunk 1 above)
Comment (mvecha) [reply code-scanning/zscaler-iac-scan.yml]: Respective commit sha is added now to the yml.

File: code-scanning/zscaler-iac-scan.yml (hunk 2 above)
Comment (mvecha) [reply code-scanning/zscaler-iac-scan.yml:18]: Required permissions are now added.

**Review by Phantsure — DISMISSED** [review]
This looks good from starter-workflow side
cc. @anuragc617 or @bishal-pdMSFT can you take a look at this as well?

**Comment by mvecha** [comment]
> This looks good from starter-workflow side
> cc. @anuragc617 or @bishal-pdMSFT can you take a look at this as well?

@anuragc617  or @bishal-pdMSFT  can you please review this pr.

File: code-scanning/zscaler-iac-scan.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+          region : 'US'
+          iac_dir : #Enter the IaC directory path from root.
+          iac_file : #Enter the IaC file path from root.
+          output_format : #(Optional) By default, the output is provided in a human readable format. However, if you require a different format, you can specify it here.
+          #To fail the build based on policy violations identified in the IaC templates, set the input value (fail_build) to true.
+          fail_build : #Enter true/false
+      #Ensure that the following step is included in order to post the scan results under the code scanning alerts section within the repository.
+      - name: Upload SARIF file
+        if: ${{ success() || failure() && (steps.zscaler-iac-scan.outputs.sarif_file_path != '') }}
+        uses: github/codeql-action/upload-sarif@v1
```
Comment (bishal-pdMSFT) [inline code-scanning/zscaler-iac-scan.yml:50]: Use v2 which is the latest version

File: code-scanning/zscaler-iac-scan.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+jobs:
+  zscaler-iac-scan:
+    permissions:
+      contents: read # for actions/checkout to fetch code
+      security-events: write # for github/codeql-action/upload-sarif to upload SARIF results
+    runs-on: ubuntu-latest
+    steps:
+      - name : Code Checkout
+        uses: actions/checkout@v2
```
Comment (bishal-pdMSFT) [inline code-scanning/zscaler-iac-scan.yml:30]: This one also is v3 now I think

File: code-scanning/zscaler-iac-scan.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+      - name : Zscaler IAC Scan
+        uses : ZscalerCWP/Zscaler-IaC-Action@8f0d8b60bd5a8f44062d444463f66f419ab71cfc
+        id : zscaler-iac-scan
+        with:
+          #Log into the Zscaler Workload Posture (ZWP) Admin Portal to begin the onboarding process.
+          #Copy the client ID and client secret key generated during the onboarding process and configure.
+          #GitHub secrets (ZSCANNER_CLIENT_ID, ZSCANNER_CLIENT_SECRET).
```
Comment (bishal-pdMSFT) [inline code-scanning/zscaler-iac-scan.yml:37]: These are pre-requisites, move them to top of file in comments.

File: code-scanning/zscaler-iac-scan.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
+#This workflow uses actions that are not certified by GitHub.
+#They are provided by a third party and are governed by
+#separate terms of service, privacy policy, and support
+#documentation.
+
+#This workflow runs the Zscaler Infrastructure as Code (IaC) Scan app,
+#which detects security misconfigurations in IaC templates and publishes the findings
+#under the code scanning alerts section within the repository.
```
Comment (bishal-pdMSFT) [inline code-scanning/zscaler-iac-scan.yml:8]: Add few more details if possible:
- Any pre-requisites
- Any link to learn more
- Any link to get support or create issue

File: code-scanning/properties/zscaler-iac-scan.properties.json [hunk 7]
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "Zscaler IaC Scan",
+  "creator": "Zscaler CWP",
+  "description": "Scan your Infrastructure as Code files using Zscaler Infrastructure as Code (IaC) Scan app",
+  "iconName": "zscaler",
```
Comment (bishal-pdMSFT) [inline code-scanning/properties/zscaler-iac-scan.properties.json:5]: Possible to have an icon only with logo and not text. The icon is shown in a small area and hence better not have any text in it.

File: code-scanning/properties/zscaler-iac-scan.properties.json [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+  "categories": ["Code Scanning","Security"]
```
Comment (bishal-pdMSFT) [inline code-scanning/properties/zscaler-iac-scan.properties.json:6]: `Security` is not needed

**Comment by mvecha** [comment]
@bishal-pdMSFT 
Updated the code. Please refer the latest code for review. 
Thanks.

File: script/validate-data/package.json [hunk 9]
```diff
@@ -14,7 +14,7 @@
     "typescript": "^3.9.2"
   },
   "dependencies": {
-    "@actions/core": "^1.2.6",
+    "@actions/core": "^1.6.0",
```
Comment (Phantsure) [inline script/validate-data/package.json:17]: Is this a needed change for this template? If not, then can you raise it separately?

**Comment by Phantsure** [comment]
@mvecha You don't have domain verification and haven't joined the partner program so we can't verify your actions. Kindly follow the steps in PR template.

File: code-scanning/properties/zscaler-iac-scan.properties.json (hunk 7 above)
Comment (Newman1515) [reply code-scanning/properties/zscaler-iac-scan.properties.json:5]: My bad

**Review by Newman1515 — CHANGES_REQUESTED** [review]
Ok

**Comment by jragula-zs** [comment]
@Phantsure The domain verification is completed for Zscaler now. Can this PR be approved?

File: script/validate-data/package.json (hunk 9 above)
Comment (Phantsure) [reply script/validate-data/package.json:17]: @mvecha Can this change be taken separately?

**Comment by Phantsure** [comment]
@jragula-zs Can you share the org for which domain verification is completed? 
The org of the action used doesn't seem to have the `verified` sign like below.
Screenshot 2022-06-13 at 5 19 39 PM

**Comment by jragula-zs** [comment]
@Phantsure The organization is verified now. Could you please check and approve?
https://github.com/ZscalerCWP

**Review by Phantsure — CHANGES_REQUESTED** [review]
If possible, can we take the update to validate script separately? Were the tests failing before the change?

File: code-scanning/properties/zscaler-iac-scan.properties.json (hunk 7 above)
Comment (Phantsure) [reply code-scanning/properties/zscaler-iac-scan.properties.json:5]: @mvecha can also take this?

**Review by Phantsure — APPROVED** [review]

**Comment by Phantsure** [comment]
@mvecha Sorry for this being here for so long. I would love to close it now. Can you merge the main again?

**Comment by jragula-zs** [comment]
> @mvecha Sorry for this being here for so long. I would love to close it now. Can you merge the main again?

@Phantsure Can you approve and merge now?

**Pull request merged**